4.6 Improve OpenShift Logging monitoring (#25159)
openshift/openshift-docs: modules/cluster-logging-dashboards-es.adoc

When Lucene performs data operations, such as a search operation, Lucene performs the operation against the index segments in the relevant index. For that purpose, each segment contains specific data structures that are loaded in the memory and mapped. Index mapping can have a significant impact on the memory used by segment data structures.

The *Logging/Elasticsearch Nodes* dashboard contains the following charts about the Elasticsearch index shards.

.Elasticsearch cluster shard status charts
[options="header"]

|===
|Metric|Description

|Cluster active shards 
|The number of active primary shards and the total number of shards, including replicas, in the cluster. If the number of shards grows higher, the cluster performance can start degrading.

|Cluster initializing shards 

> |The number of non-active shards in the cluster. A cluster typically has non–active shards for short periods. A growing number of non–active shards over longer periods could indicate a problem.

Reviewer: If you have another sentence about why a cluster only has non-active shards for a short period or the problem, it might be useful to add.

Each Elasticsearch node has a finite amount of resources that can be used to process tasks. When all the resources are being used and Elasticsearch attempts to perform a new task, Elasticsearch put the tasks into a queue until some resources become available. 

The *Logging/Elasticsearch Nodes* dashboard contains the following charts about resource usage for a selected node and the number of tasks waiting in the Elasticsearch queue.

.Elasticsearch node metric charts
[options="header"]
|===
|Metric|Description

|ThreadPool tasks
|The number of waiting tasks in individual queues, shown by task type. A long–term accumulation of tasks in any queue could indicate node resource shortages or some other problem.

|CPU usage 
|The amount of CPU being used by the selected Elasticsearch node as a percentage of the total CPU allocated to the host container.

|Memory usage 
|The amount of memory being used by the selected Elasticsearch node.

|Disk usage 
|The total disk space being used for index data and metadata on the selected Elasticsearch node. 

|Documents indexing rate 
|The rate that documents are indexed on the selected Elasticsearch node.

|Indexing latency 

> |The time taken to index the documents on the selected Elasticsearch node. Indexing latency can be affected by many factors, such as JVM Heap memory and overall load. A growing trend indicates a resource capacity shortage in the instance.

Reviewer: ```suggestion
|The time taken to index the documents on the selected Elasticsearch node. Indexing latency can be affected by many factors, such as JVM Heap memory and overall load. A growing latency trend indicates a resource capacity shortage in the instance.
```

|Search rate 

> |The number of search requests executed on the selected Elasticsearch node.

Reviewer: ```suggestion
|The number of search requests run on the selected Elasticsearch node.
```

Unless execute is really the right Elasticsearch term.

|Search latency  
|The time taken to complete search requests on the selected Elasticsearch node. Search latency can be affected by many factors. A growing trend indicates a resource capacity shortage in the instance.

|Documents count (with replicas) 
|The number of Elasticsearch documents stored on the selected Elasticsearch node, including documents stored in both the primary shards and replica shards that are allocated on the node.

|Documents deleting rate 

> |The number of Elasticsearch documents being deleted from any of the index shards allocated to the selected Elasticsearch node.

Reviewer: ```suggestion
|The number of Elasticsearch documents being deleted from any of the index shards that are allocated to the selected Elasticsearch node.
```

|Documents merging rate 

> |The number of Elasticsearch documents being merged in any of index shards allocated to the selected Elasticsearch node.

Reviewer: ```suggestion
|The number of Elasticsearch documents being merged in any of index shards that are allocated to the selected Elasticsearch node.
```

|===

Elasticsearch node fielddata::

> _Fielddata_ is a query-time in-memory data structure for text fields that is built by Elasticsearch at the index-segment level and kept in the JVM Heap. Because fielddata building is an expensive operation, Elasticsearch caches the fielddata structures. Elasticsearch can evict a fielddata cache when the underlying index segment is deleted or merged, or if there is not enough JVM HEAP memory for all the fielddata caches. 

Reviewer: I don't think the first sentence is going to translate well. Can you use simpler language?
```suggestion
_Fielddata_ is a query-time in-memory data structure for text fields that is built by Elasticsearch at the index-segment level and kept in the JVM Heap. Because fielddata building is an expensive operation, Elasticsearch caches the fielddata structures. Elasticsearch can evict a fielddata cache when the underlying index segment is deleted or merged or if there is not enough JVM HEAP memory for all the fielddata caches. 
```

.Elasticsearch node fielddata charts
[options="header"]
|===
|Metric|Description

|Fielddata memory size 
|The amount of JVM Heap used for the fielddata cache on the selected Elasticsearch node.

|Fielddata evictions 
|The number of fielddata structures that were deleted from the selected Elasticsearch node. 

|===

Elasticsearch node query cache::

If the data stored in the index does not change, search query results are cached in a node-level query cache for reuse by Elasticsearch. 

The *Logging/Elasticsearch Nodes* dashboard contains the following charts about the Elasticsearch node query cache.

.Elasticsearch node query charts
[options="header"]
|===
|Metric|Description

|Query cache size 
|The total amount of memory used for the query cache for all the shards allocated to the selected Elasticsearch node.

|Query cache evictions 
|The number of query cache evictions on the selected Elasticsearch node.

|Query cache hits 
|The number of query cache hits on the selected Elasticsearch node.

|Query cache misses 
|The number of query cache misses on the selected Elasticsearch node.

|===

Elasticsearch index throttling::

> When indexing documents, Elasticsearch stores the documents in index segments, which are physical representations of the data. At the same time, Elasticsearch periodically merges smaller segments into a larger segment as a way to optimize resource use. If the indexing is faster then the ability to merge segments, the merge process gets behind. To prevent this situation, Elasticsearch throttles indexing, typically by reducing the number of threads allocated to indexing down to a single thread. 

Reviewer: ```suggestion
When indexing documents, Elasticsearch stores the documents in index segments, which are physical representations of the data. At the same time, Elasticsearch periodically merges smaller segments into a larger segment as a way to optimize resource use. If the indexing is faster then the ability to merge segments, the merge process does not complete quickly enough <and why this is a problem>. To prevent this situation, Elasticsearch throttles indexing, typically by reducing the number of threads allocated to indexing down to a single thread. 
```

The *Logging/Elasticsearch Nodes* dashboard contains the following charts about Elasticsearch index throttling.

.Index throttling charts
[options="header"]
|===
|Metric|Description

|Indexing throttling 
|The amount of time that Elasticsearch has been throttling the indexing operations on the selected Elasticsearch node.

|Merging throttling 
|The amount of time that Elasticsearch has been throttling the segment merge operations on the selected Elasticsearch node.

|===

Node JVM Heap statistics::

The *Logging/Elasticsearch Nodes* dashboard contains the following charts about JVM Heap operations.

.JVM Heap statistic charts
[options="header"]
|===
|Metric|Description

|Heap used 

> |The amount of the total allocated JVM Heap space being used on the selected Elasticsearch node.

Reviewer: ```suggestion
|The amount of the total allocated JVM Heap space that is used on the selected Elasticsearch node.
```

|GC count 

> |The number of garbage collection operations that have been executed on the selected Elasticsearch node, by old and young garbage collection.

Reviewer: ```suggestion
|The number of garbage collection operations that have been run on the selected Elasticsearch node, by old and young garbage collection.
```
I'm afraid to ask, but what are old and young garbage collection?

|GC time 

> |The amount of time that the JVM spent executing garbage collection operations on the selected Elasticsearch node, by old and young garbage collection.

Reviewer: ```suggestion
|The amount of time that the JVM spent completing garbage collection operations on the selected Elasticsearch node, by old and young garbage collection.
```

// Module included in the following assemblies:
//
// * logging/cluster-logging-dashboards.adoc

[id="cluster-logging-dashboards-es_{context}"]
= About the Logging/Elastisearch nodes dashboard  

The *Logging/Elasticsearch nodes* dashboard contains charts that show details about your Elasticsearch instance, many at node-level, for further diagnostics.

Elasticsearch status::  

The {product-title} *Logging/Elasticsearch nodes* dashboard contains the following charts about the status of your Elasticsearch instance.

.Elasticsearch status fields
[options="header"]
|===
|Metric|Description

|Cluster status 
a|The cluster health status during the selected time period. Each time series represents one of possible states: green, yellow or red. A value of `0` indicates the state is green.

> * 0 - Indicates the cluster is in green status, which means that all shards are allocated. 

Reviewer 2: We have 'green' words now.
